### ISSUE TITLE
Outsource relevant code to package instead of starter-kit

### ISSUE THREAD

**Issue by bkniffler** [ctrlplusb/react-universally#140]
Hi, 
This starter is awesome! Did you consider making this a package that can be installed to a project instead of a starter-kit? 

There is a few pros by going dependency:
- Keep up to date more easily as this package would reference the relevant deps (webpack, babel, etc.)
- Better app code separation from runtime/deployment
- Plugin less, preact etc more easily by installing something like react-universally-preact, react-universally-less

The drawback would be a more complicated codebase for this package.

Is this something you'd be interested in?

**Comment by ctrlplusb** [comment]
Hi @bkniffler!

Thanks man! I am really glad you like it. :)

Moving it towards the model you describe is definitely something that I would be interested in, but I am first trying it out against a few production projects.  I've found that this has been exposing a lot of the pain points with the current configuration, and has allowed me to make some useful adjustments to it. :)

I agree that it will be a massive undertaking with a lot more effort in managing change and ensuring that semantic release is upheld extremely strictly, but the benefits of doing so are definitely clear.

So to sum up, once I have a few successful production projects and feel that the project is fairly settled I would be happy to move towards this.  I also hope to have a list of collaborators added to the project by that time to help guide the project in right direction. :)

**Comment by bkniffler** [comment]
Thanks for the quick response. I see your point that it could make it harder to progressively improve the applications foundation, then again you have less redundancy.

I still started experimenting with the idea, here is the branch: https://github.com/bkniffler/react-universally. It is never too early to check viability of useful enhancements :)

In the demo folder you'll find working example, just yarn install inside. Notice there are only two dependencies in the package.json.

```
"dependencies": {
    "@bkniffler/react-universally": "^7.2.2"
  },
  "devDependencies": {
    "@bkniffler/react-universally-dev": "^7.2.2"
  }
```

I didn't have to change **too** much, just a few adjustments to paths and adding resolve.alias to get the app directory and some more stuff. I know there is still things to improve (like not hard-coding the app directory and allowing to extend the express server), but I'm really positive on this. Especially since I can setup a symlink to react-universally from the demo projects node_modules to ease the coding experience.

**Comment by ctrlplusb** [comment]
Fantastic stuff @bkniffler - that was faster than expected!

I'll review this when I can, and perhaps we can work together on this. :)

**Comment by ctrlplusb** [comment]
I have given this a lot of thought and it's definitely something I would like to pursue. :)

I am doing a cleanup parse and refactor on the project, after which I want to get the toolchain separated (possibly using `tension`), and then will look towards the creation of a cli.

I'll keep you posted, I would appreciate your help. :)

**Comment by bkniffler** [comment]
Yeah, sure. I'm actively using this approach here: https://github.com/olymp/olymp
Everything works well, even devmode with the package linked. I've tried to keep the boilerplate code of this repository as much seperated from other stuff as possible, so 

Things that I needed to change:
- decouple the react app folder (using a configurable path, I'm using a webpack alias called `app_alias`)
- allow to plug-in server-side code (I'm using a webpack alias again, called `server_alias`)
- I added some entries to the CSP code, so this will need to be more configurable
- Added some babel plugins manually, this will need to be configurable
- Create a olymp package that has everything from react-universally and an olymp-dev package that only has the devDependencies of react-universally. olymp-dev will then be a devDependency of the final project.
- Take care of duplicate react/react-dom instances and other packages if you link react-universally (as soon as it is possible to use as a dependency), using webpack aliases works well here.
- Think about either statically compiling the current src folder of react-universally, or to include `node_modules/react-universally/src` into the webpack config
- Hot reloading works good with using the `app_alias`

I didn't look into tension too much, but I'd love the webpack configuration to be very simple to extend, the middleware approach could work really well here. Each middleware would just extend it more and more.

Also I'd love if the configuration could be reconsidered. Some problems:
- Can't assign process.env.PORT (which services like heroku set) to SERVER_PORT setting that react-universally expects.
- I don't like to be forced to have this setup on each projects with the same variables (BUNDLE_OUTPUT_PATH, BUNDLE_ASSETS_FILENAME, etc). Why not just a .js file that returns an object and react-universally would add sane defaults, even environment awareness wouldn't be a problem since we could use process.env.NODE_ENV, we could assign `SERVER_PORT: process.env.PORT`, also inheriting a base config that people could define somewhere wouldn't be a problem with Object.assign.

**Comment by ctrlplusb** [comment]
Thats awesome thanks for those details!  I will make some time to go through what you have.

In regards to your other concerns.  I completely agree. I think I have confused configuration over project meta information.  Like you say BUNDLE_OUTPUT_PATH etc hardly ever change.  They shouldn't really be exposed as configuration options, but rather be a higher level project meta configuration which can be manually overridden in code if required.  I plan to make a lot of changes around this soon.

**Comment by bkniffler** [comment]
Just to be clear, I'm very happy that you want to make it possible to have react-universally as a dependency instead of boilerplate and I'm going to help if you need me :)

**Comment by ctrlplusb** [comment]
Closing in favour of #201

**Comment by bkniffler** [comment]
Hey @ctrlplusb, just wanted to let you know I've incorporated the latest react-universally into my package. Also, I've cleaned it up, so this truly is a dependable version of the latest react-universally.

You can see the example here: https://github.com/olymp/universally/tree/master/example

Some infos:
- due to the bin/universally file, `universally dev` or `universally build` commands are possible
- a file `universally.config.js` is automatically picked up from your project root and used to extend react-universally config
- modules (universally-apollo, universally-less) are possible and exist, you can just modify your `universally.config.js`
```
import less from 'universally-dev-less';

export default config => less({
  ...config,
  bundlesSharedSrcPath: './app',
  port: process.env.PORT || 3010,
  clientDevServerPort: process.env.PORT ? process.env.PORT + 1 : 3011
});
```

This is really a neat alternative to create-react-app and next.js, since its fully extendable (due to your excellent config system) and it comes with plenty of features. It might be the best easy way to get a serious project started.

Are you interested in making this a feature (via branch or in master) in this repo? Or would you rather like this to be completely separated?

**Comment by ctrlplusb** [comment]
Dude! Wow man this looks great.  I want to look at this a bit more in detail before I respond.

Sorry, swamped with client work at the moment.
